include/aspect/material_model/viscoelastic.h

```diff
@@ -0,0 +1,232 @@
+/*
+  Copyright (C) 2014 - 2017 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef _aspect_material_model_viscoelastic_h
+#define _aspect_material_model_viscoelastic_h
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <deal.II/base/parameter_handler.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    /**
+     * A material model which is intended for simple linear viscoelastic
+     * problems with multiple compositional fields.
+     *
+     * The viscoelastic constitutive relationship and implementation follows
+     * a method commonly used in the Geodynamics community, where only the
+     * elastic shear strength of rocks (e.g., shear modulus) is considered.
+     * The material model directly follows the work flow and terminology
+     * of Moresi et al. 2003 (J. Comp. Phys., v. 184, p. 476-497) equations
+     * 23-32, which is commonly referred to in proceeding geodynamic
+     * publications. However, a notable difference between this material
+     * model and that of previous work is the use of compositional fields,
+     * rather than tracers, to track and advect the stress tensor. The
+     * material model will be updated when an option to track and advect
+     * stresses with tracers is implemented.
+     *
+     * The first 3 (2D) or 6 (3D) compositional fields are used to track
+     * individual components of the viscoelastic stress tensor, while additional
+     * fields represent distinct rock types.
```

Comment:
I agree the naming convention is much safer and I actually meant to do this before updating the pull request. The only advantage to the current method is that it only requires two lines of codes, but it certainly could lead to all sorts of issues. However, the user will still have to be quite careful when assigning properties to a large number of fields in 3D (>10 in some cases?). Would it be worth requiring both a specific naming scheme and ordering as an extra 'safety' feature?

Reply:
For now I am fine either way. I can imagine another plugin requiring that its special properties have to be fields 0..2, then we have a problem. But that could be resolved when it finally crops up. Make sure you require the existence of the names though, otherwise allowing for variable ordering but requiring the names in the future will break compatibility.